### PR TITLE
feat(flink): implement UDF support for the backend

### LINKED ISSUE #8108: feat(flink): implement UDF support for the backend
### Is your feature request related to a problem?

* I want to be able to execute the same UDFs that I can against other backends against the Flink backend.
* I want to be able to define custom aggregations.
* I want to use external libraries in UDFs.
* I want to load external resources (e.g. models) in UDFs.

##### Definition

Ibis supports four scalar backend-agnostic UDF definitions. For the full documentation, see https://ibis-project.org/reference/scalar-udfs.

* `builtin`:  is built in to the backend
* `pandas`: accepts pandas Series as inputs
* `pyarrow`: accepts PyArrow Arrays as inputs
* `python`: accepts Python scalars as inputs

Ibis also supports built-in aggregate UDFs (https://ibis-project.org/how-to/extending/builtin.html#aggregate-functions), but doesn’t expose an API for defining custom aggregate functions.

##### Registration

* Scalar UDFs are registered in pre-execute hooks: https://github.com/ibis-project/ibis/blob/7.2.0/ibis/backends/base/__init__.py#L1009
* Some backends also register vectorized UDFs in the _register_udfs method: https://github.com/ibis-project/ibis/blob/7.2.0/ibis/backends/datafusion/__init__.py#L146-L148
* Direct execution backends (e.g. [pandas](https://github.com/ibis-project/ibis/blob/7.2.0/ibis/backends/pandas/udf.py), [dask](https://github.com/ibis-project/ibis/blob/7.2.0/ibis/backends/dask/udf.py), polars) use other methods
* Some backends (e.g. [sqlite](https://github.com/ibis-project/ibis/blob/7.2.0/ibis/backends/sqlite/udf.py#L31-L53)) register custom UDFs with the connection

##### Translation

https://github.com/ibis-project/ibis/blob/7.2.0/ibis/backends/base/sql/__init__.py#L283-L294

#### UDF support in the PyFlink Table API

The PyFlink Table API supports:

* [Scalar functions](https://nightlies.apache.org/flink/flink-docs-release-1.18/docs/dev/python/table/udfs/python_udfs/#scalar-functions): takes zero, one, or multiple scalar values as input parameters and returns a single value
* [Table functions](https://nightlies.apache.org/flink/flink-docs-release-1.18/docs/dev/python/table/udfs/python_udfs/#table-functions): takes zero, one, or multiple scalar values as input parameters and returns an arbitrary number of rows consisting of one or more fields
* [Aggregate functions](https://nightlies.apache.org/flink/flink-docs-release-1.18/docs/dev/python/table/udfs/python_udfs/#aggregate-functions): maps scalar values of multiple rows to a new scalar value
* [Table aggregate functions](https://nightlies.apache.org/flink/flink-docs-release-1.18/docs/dev/python/table/udfs/python_udfs/#table-aggregate-functions): maps scalar values of multiple rows to an arbitrary number of rows consisting of one or more fields
* [Vectorized scalar functions](https://nightlies.apache.org/flink/flink-docs-release-1.18/docs/dev/python/table/udfs/vectorized_python_udfs/#vectorized-scalar-functions): accepts pandas Series as inputs and returns pandas Series of the same length as outputs
* [Vectorized aggregate functions](https://nightlies.apache.org/flink/flink-docs-release-1.18/docs/dev/python/table/udfs/vectorized_python_udfs/#vectorized-aggregate-functions): takes one or more pandas Series as inputs and returns one scalar value as the output

Other deployment considerations (e.g. using external libraries and resources): https://nightlies.apache.org/flink/flink-docs-release-1.18/docs/dev/python/table/udfs/overview/

#### Mapping the Ibis and PyFlink Table APIs

Ibis | PyFlink Table API
-- | --
`ibis.udf.scalar.builtin` | Call an already-registered UDF (or a function not exposed by the Ibis API)
`ibis.udf.scalar.pandas` (`ops.ElementWiseVectorizedUDF`) | Register and call a vectorized scalar function
`ibis.udf.scalar.pandas` (`ops.ReductionVectorizedUDF`) | Register and call a vectorized aggregate function
`ibis.udf.scalar.pyarrow` | Not supported
`ibis.udf.scalar.python` | Call and register a scalar function
Not supported[^_1] | Table functions
`ibis.udf.agg.builtin` | Call an already-registered UDF (or a function not exposed by the Ibis API)
Custom aggregation functions are not currently supported | Aggregate functions
Custom aggregation functions are not currently supported | Table aggregate functions

#### Implementation

In order of ease:

* Built-in UDF support should be straightforward, if it doesn’t already work, since it doesn’t require definition/registration.
* Scalar Python and pandas UDFs will require registration, but the DataFusion backend looks like a reasonable example of how to do that.[^_2]
* Custom aggregation will require Ibis API design, and the resulting syntax should be useful across multiple backends. Is this already part of the upcoming milestone around UDFs?

[^_1]: I'm not 100% confident about this.
[^_2]: Question for somebody with more experience implementing Ibis UDFs: Is it a good example, or does the Postgres backend provide a better template?

### What version of ibis are you running?

dev

### What backend(s) are you using, if any?

Flink

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

### PR DESCRIPTION
**Pull request by deepyaman**
## Description of changes

* Add UDF translation rules to the Flink backend (copied from the base SQL backend, because Flink doesn't extend it)
  * Maybe this could be cleaner/not require copy-pasting?
* Add a test for a built-in scalar UDF
* Add a test for a built-in aggregation UDF
* Implement scalar Python UDFs
* Implement scalar pandas UDFs

## Issues closed

Resolves #8108

### CODE REVIEW

File: ibis/backends/flink/tests/test_udf.py [hunk 1]
```diff
@@ -0,0 +1,13 @@
+from __future__ import annotations
+
+from ibis import udf
+
+
+def test_builtin_scalar_udf(con):
+    @udf.scalar.builtin
+    def parse_url(string1: str, string2: str) -> str:
```
Comment (deepyaman) [inline ibis/backends/flink/tests/test_udf.py:8]: I actually also wanted to add a test for the 3-argument variant, but I couldn't find a "reasonable" way to do this. It seems like both UDFs would need to have the name `parse_url`, so that would mean I'd need to construct them in a way that doesn't overwrite each other. There's no way to make an optional argument.

In the end, I didn't add this workaround in the tests, because that's not how a user would use it; would I be correct to assume a user can't have both `PARSE_URL(string1, string2)` and `PARSE_URL(string1, string2, string3)` in an Ibis expression?

**Comment by cpcloud** [comment]
Any chance we can hold off on this for the 8.0 release and target `the-epic-split`. Flink isn't yet ported (working on it now) so there's nothing to code against for the moment.

File: ibis/backends/flink/tests/test_udf.py (hunk 1 above)
Comment (cpcloud) [reply ibis/backends/flink/tests/test_udf.py:8]: We don't have a way of overloading functions at the moment, so yeah a user would have to encode these as separate functions:

```python
@udf.scalar.builtin(name="parse_url")
def parse_url2(a, b):
	...

@udf.scalar.builtin(name="parse_url")
def parse_url3(a, b, c):
	...
```

Comment (deepyaman) [reply ibis/backends/flink/tests/test_udf.py:8]: Ah, I didn't see the `name` argument; that works then!

**Comment by deepyaman** [comment]
> Any chance we can hold off on this for the 8.0 release and target `the-epic-split`? Flink isn't yet ported (working on it now) so there's nothing to code against for the moment.

Sure! I don't think there's any rush on this.

Maybe I'll just keep this branch for now, and then will make whatever changes and/or rebase on top of the new branch, raising a new PR if need be?

**Comment by deepyaman** [comment]
UDF tests pass locally:

```
(ibis-dev) deepyaman@Deepyamans-Air ibis % poetry run pytest --junitxml=junit.xml --cov=ibis --cov-report=xml:coverage.xml -m flink --randomly-dont-reorganize ibis/backends/tests/test_udf.py 
============================================================================================= test session starts ==============================================================================================
platform darwin -- Python 3.10.13, pytest-7.4.3, pluggy-1.3.0
Using --randomly-seed=3454285181
benchmark: 4.0.0 (defaults: timer=time.perf_counter disable_gc=False min_rounds=5 min_time=0.000005 max_time=1.0 calibration_precision=10 warmup=False warmup_iterations=100000)
rootdir: /Users/deepyaman/github/claypotai/ibis
configfile: pyproject.toml
plugins: repeat-0.9.2, hypothesis-6.88.1, snapshot-0.0.0, pytest_httpserver-1.0.8, randomly-3.15.0, cov-4.1.0, clarity-1.0.1, mock-3.12.0, xdist-3.3.1, benchmark-4.0.0
collected 100 items / 95 deselected / 5 selected                                                                                                                                                               

ibis/backends/tests/test_udf.py Exception ignored in: <_io.FileIO name=26 mode='wb' closefd=True>
Traceback (most recent call last):
  File "/opt/miniconda3/envs/ibis-dev/lib/python3.10/site-packages/_pytest/unraisableexception.py", line 59, in __exit__
    del self.unraisable
ResourceWarning: unclosed file <_io.BufferedWriter name=26>
Exception ignored in: <_io.FileIO name=27 mode='rb' closefd=True>
Traceback (most recent call last):
  File "/opt/miniconda3/envs/ibis-dev/lib/python3.10/site-packages/_pytest/unraisableexception.py", line 59, in __exit__
    del self.unraisable
ResourceWarning: unclosed file <_io.BufferedReader name=27>
.xx.x                                                                                                                                                                    [100%]

--------------------------------------------------------------------- generated xml file: /Users/deepyaman/github/claypotai/ibis/junit.xml ---------------------------------------------------------------------

--------- coverage: platform darwin, python 3.10.13-final-0 ----------
Coverage XML written to file coverage.xml

================================================================================= 2 passed, 95 deselected, 3 xfailed in 24.28s =================================================================================
```

File: docker/flink/Dockerfile [hunk 2]
```diff
@@ -1,3 +1,12 @@
...
 # ibis-flink requires PyFlink dependency
 RUN wget -nv -P $FLINK_HOME/lib/ https://repo1.maven.org/maven2/org/apache/flink/flink-python/1.18.1/flink-python-1.18.1.jar
+
+# install python3 and pip3
+RUN apt-get update -y && \
+apt-get install -y python3 python3-pip python3-dev && rm -rf /var/lib/apt/lists/*
+RUN ln -s /usr/bin/python3 /usr/bin/python
+
+# install PyFlink
+RUN pip3 install apache-flink==1.18.1
```
Comment (deepyaman) [inline docker/flink/Dockerfile:12]: Actually, let me check if this bit is necessary.

File: ibis/backends/flink/tests/test_udf.py [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+
+def test_builtin_agg_udf(con):
+    @udf.agg.builtin
+    def json_arrayagg(a) -> str:
+        """Glom together some JSON."""
+
+    ft = con.tables.functional_alltypes[:5]
+    expr = json_arrayagg(ft.string_col)
+    result = expr.execute()
+    expected = json.dumps(list(map(str, range(5))), separators=",:")
```
Comment (cpcloud) [inline ibis/backends/flink/tests/test_udf.py:26]:
```suggestion
    expected = '["1","2","3","4","5"]'
```

Just to make it a bit more obvious what the JSON looks like :)

Comment (deepyaman) [reply ibis/backends/flink/tests/test_udf.py:26]: Sounds good; I'll update it here and wherever I copied it from (`mysql`, I think).

File: docker/flink/Dockerfile (hunk 2 above)
Comment (cpcloud) [reply docker/flink/Dockerfile:12]: Did this end up being necessary?

File: ibis/backends/flink/__init__.py [hunk 4]
```diff
@@ -301,6 +304,36 @@ def version(self) -> str:
 
         return pyflink.version.__version__
 
+    def _register_udfs(self, expr: ir.Expr) -> None:
```
Comment (cpcloud) [inline ibis/backends/flink/__init__.py:307]: This logic already exists under `_compile_udf` in the `SQLGlotBackend` base class. I know the name isn't really the right one, but can you reuse it here instead of reimplementing very similar logic?

Comment (cpcloud) [reply ibis/backends/flink/__init__.py:307]: Sorry, I was looking at the wrong backend, scratch what I said.

File: docker/flink/Dockerfile (hunk 2 above)
Comment (deepyaman) [reply docker/flink/Dockerfile:12]: It is.

Comment (cpcloud) [reply docker/flink/Dockerfile:12]: Not blocking the PR, but is this dependency on the server side inevitable for the foreseeable future?

**Review by cpcloud — APPROVED** [review]
🚢 it!

**Comment by gforsyth** [comment]
#8398 is in now

**Comment by deepyaman** [comment]
> @deepyaman Can we wait for #8398 to be merged? You'll be able to delete some boilerplate in this PR once that one's in.

Sounds good! I'll update this afternoon.

File: docker/flink/Dockerfile (hunk 2 above)
Comment (deepyaman) [reply docker/flink/Dockerfile:12]: I believe so. I think this is similar to how something like Spark also works, where you have to make any Python libraries/files needed by UDFs available to the cluster. Otherwise, you have to submit to the cluster when running, using something like https://nightlies.apache.org/flink/flink-docs-master/docs/dev/python/python_config/#python-options

File: ibis/backends/duckdb/__init__.py
Comment (cpcloud) [inline ibis/backends/duckdb/__init__.py:1]: You should be able to revert the changes in this file.

File: ibis/backends/duckdb/__init__.py [hunk 5]
```diff
@@ -1463,6 +1461,12 @@ def register_udf(con):
     _compile_python_udf = _compile_udf
     _compile_pyarrow_udf = _compile_udf
 
+    def _compile_builtin_udf(self, udf_node: ops.ScalarUDF) -> None:
+        """No op."""
+
+    def _compile_pandas_udf(self, _: ops.ScalarUDF) -> None:
+        raise NotImplementedError("DuckDB doesn't support pandas UDFs")
+
```
Comment (deepyaman) [inline ibis/backends/duckdb/__init__.py:1469]:
```suggestion
```

File: ibis/backends/duckdb/__init__.py
Comment (deepyaman) [reply ibis/backends/duckdb/__init__.py:1]: I think the other changes are still valid, but I've reverted the fallback behaviors.

**Comment by cpcloud** [comment]
Thanks!
